src/freenas/usr/local/lib/middlewared_truenas/plugins/failover_/event_linux.py

```diff
@@ -0,0 +1,713 @@
+            zpool_cache_saved_mtime = os.stat(self.ZPOOL_CACHE_FILE_SAVED).st_mtime
+            if zpool_cache_mtime > zpool_cache_saved_mtime:
+                with contextlib.suppress(Exception):
+                    shutil.copy2(self.ZPOOL_CACHE_FILE, self.ZPOOL_CACHE_FILE_SAVED)
+
+        # set the progress to IMPORTING
+        job.set_progress(None, description='IMPORTING')
+
+        failed = []
+        for vol in fobj['volumes']:
+            logger.info('Importing %s', vol['name'])
+
+            # import the zpool(s)
+            try:
+                self.run_call(
+                    'zfs.pool.import_pool',
+                    vol['guid'],
+                    {
+                        'altroot': '/mnt',
+                        'cachefile': self.ZPOOL_CACHE_FILE,
+                    }
+                )
+            except Exception as e:
+                vol['error'] = str(e)
+                failed.append(vol)
+                continue
+
+            # try to unlock the zfs datasets (if any)
+            unlock_job = self.run_call('failover.unlock_zfs_datasets', vol["name"])
+            unlock_job.wait_sync()
+            if unlock_job.error:
+                logger.error(f'Error unlocking ZFS encrypted datasets: {unlock_job.error}')
+            elif unlock_job.result['failed']:
+                logger.error('Failed to unlock %s ZFS encrypted dataset(s)', ','.join(unlock_job.result['failed']))
+
+        # if we fail to import all zpools then alert the user because nothing
+        # is going to work at this point
+        if len(failed) == len(fobj['volumes']):
+            for i in failed:
+                logger.error(
+                    'Failed to import volume with name "%s" with guid "%s" '
+                    'with error "%s"', failed['name'], failed['guid'], failed['error'],
+                )
+
+            logger.error('All volumes failed to import!')
+            job.set_progress(None, description='ERROR')
+            raise AllZpoolsFailedToImport()
+
+        # if we fail to import any of the zpools then alert the user but continue the process
+        elif len(failed):
+            for i in failed:
+                logger.error(
+                    'Failed to import volume with name "%s" with guid "%s" '
+                    'with error "%s"', failed['name'], failed['guid'], failed['error'],
+                )
+                logger.error(
+                    'However, other zpools imported so the failover process continued.'
+                )
+
+        logger.info('Volume imports complete.')
+
+        # need to make sure failover status is updated in the middleware cache
+        logger.info('Refreshing failover status')
+        self.run_call('failover.status_refresh')
+
+        # this enables all necessary services that have been enabled by the user
+        logger.info('Enabling necessary services')
+        self.run_call('etc.generate', 'rc')
```

Comment:
Is this necessary for Linux?

Reply:
As far as I know it is, it actually calls `systemd.py` which will run `systemctl enable "service"` if it isn't enabled already.